feat(components): modify autocomplete to scroll list up and down

dmcrorieGIT opened the pull request:
<!--
  Atlantis uses Conventional Commits to track versions.
  Pull request titles should follow the following format.

  <TYPE>(<optional SCOPE>): <conditionally BREAKING CHANGE:> <description>

  eg.
    fix(SCOPE): stop graphite breaking when too much pressure applied — Patch Release
    feat(SCOPE): add 'graphiteWidth' option — (Minor) Feature Release
    feat(SCOPE): BREAKING CHANGE: remove graphiteWidth option — (Major) Breaking Release

  TYPE should consist of:
    - fix: a commit of the type fix patches a bug in your codebase
    - feat: a commit of the type feat introduces a new feature to the codebase
    - docs: documentation only changes
    - build: improvements to the build system
    - refactor: a change that neither fixes a bug nor introduces a feature
    - chore: other changes that don't modify src or test files

  SCOPE should be one of:
    - components
    - design
    - eslint
    - generators
    - hooks
    - stylelint


  If your pull request introduces a breaking change please append `BREAKING CHANGE:` following type / scope.

  Further Reading:
    - https://www.conventionalcommits.org
    - https://github.com/commitizen/conventional-commit-types/blob/master/index.json
-->

## Motivations

Autocomplete highlighting behaviour must be changed as highlighted option falls out of view when scrolling too far down or too far up

### Changed

- When list item falls out of view on menu list, the list will scroll to keep that item in view
![autocomplete-scrolling](https://user-images.githubusercontent.com/34727471/93142564-220f1900-f69b-11ea-90c7-bfbcafebdf8b.gif)


## Testing

- [ ] change the `options` in the mdx file for `Autocomplete` to include a bunch of options with labels that start with the same letter
- [ ] In the autocomplete playground, type in that first letter that most of your options begin with
- [ ] Start scrolling down with the arrow keys
- [ ] When you get to the bottom of the list and keep going, ensure that the list scrolls so that the highlighted item never falls out of view
- [ ] Ensure the same for scrolling up

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

![Random photo of Atlantis](https://loremflickr.com/672/400/atlantis)

willcodeforcoffee commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -44,7 +47,17 @@ export function Menu({
-    <div className={optionMenuClass}>
+    <div
+      className={optionMenuClass}
+      style={{ scrollBehavior: "smooth" }}
```
Can you add this to the `optionMenuClass` instead of adding a style attribute?

dmcrorieGIT replied:
Yup, added the style into the css file!

willcodeforcoffee commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -44,7 +47,17 @@ export function Menu({
+      className={optionMenuClass}
+      style={{ scrollBehavior: "smooth" }}
+      ref={ref => {
+        if (!menuDiv.current) {
+          const stuff = ref as HTMLDivElement;
+          menuDiv.current = stuff;
+        }
+        return menuDiv.current;
+      }}
```
I don't understand why you need to build the ref this way? Can'y you just set `ref={menuDiv}`?

dmcrorieGIT replied:
@willcodeforcoffee It works, but typescript doesn't seem to like that

![Screen Shot 2020-09-14 at 1 46 07 PM](https://user-images.githubusercontent.com/34727471/93136392-b32cc280-f690-11ea-9870-963194ab0831.png)

willcodeforcoffee commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -63,7 +76,16 @@ export function Menu({
-            <div className={styles.icon}>
+            <div
+              className={styles.icon}
+              ref={ref => {
+                if (!menuItemRefs.current[index]) {
+                  const newMenuRef = ref as HTMLDivElement;
+                  menuItemRefs.current[index] = newMenuRef;
+                }
+                return menuItemRefs.current[index];
+              }}
```
This looks really similar to the above code, too. Maybe it should be refactored into a method?

dmcrorieGIT replied:
Good idea, refactored into it's own method!

seanhealy commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -29,6 +30,8 @@ export function Menu({
   const optionMenuClass = classnames(styles.options, {
     [styles.visible]: visible,
   });
+  const menuItemRefs = useRef<HTMLDivElement[]>([]);
```
I don't think we need this ref. We can use `queryBySelector` getting the n-th child in the scrolling function to accomplish the same thing.

dmcrorieGIT replied:
Good idea!! Now using query by classname instead of setting refs 😄

seanhealy commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -19,6 +19,7 @@ interface MenuProps {
   onOptionSelect(chosenOption: Option): void;
 }
 
+// eslint-disable-next-line max-statements
```
How many statements are we at? We should avoid doing this if we can. I think we have a few opportunities to simplify the logic in here.

dmcrorieGIT replied:
Yup, totally! Simplified logic and removed max statement es-lint disable

seanhealy commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -87,6 +100,24 @@ export function Menu({
     </div>
   );
 
+  function setRef(
```
I'm not following why we need this method?

dmcrorieGIT replied:
We were set the refs in a DRY way, but now that we only need the ref for the menu, we don't need this anymore! Killed this method ⚔️

seanhealy commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -99,15 +130,53 @@ export function Menu({
+      }
     });
 
+    function scrollDownIfTooLow(
```
I think we should be able to generalize this into a single method with `scrollUpIfTooHigh`.

dmcrorieGIT replied:
Totally, DRY'd up these methods into one!

seanhealy reviewed (changes requested):
Hey, got some quick comments in to get y'all unblocked. Hope they make sense. Sorry they are kind of right to the point. Happy to talk about any of them or look at them together as well.

dmcrorieGIT commented:
@seanhealy Thanks for the feedback!! I've addressed all comments, could you take another look? 😄

seanhealy commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -130,50 +101,41 @@ export function Menu({
+    function scrollMenuIfItemNotInView(
       menuDivElement: HTMLDivElement,
-      itemDivElement: HTMLDivElement,
+      direction: string,
```
```suggestion
      direction: "up" | "down",
```

dmcrorieGIT replied:
Yup, it makes sense to constrict the type on this parameter! Done 😎

seanhealy commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -130,50 +101,41 @@ export function Menu({
+      direction: string,
     ) {
-      const tempDiv = itemDivElement.parentElement as HTMLDivElement;
+      const tempDiv = menuDivElement.getElementsByClassName(
```
I _think_ you may be better doing something like:

```ts
const option = menuDivElement.querySelector(`${CLASSNAME}:nth-child(${highlightedIndex})`)
```

This will push all the querying into the browser so it will be more efficient about it and we don't need an array here.

dmcrorieGIT replied:
Yup, that's a better way of query selecting; I'll do that!

seanhealy commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -50,7 +48,7 @@ export function Menu({
       className={optionMenuClass}
       ref={element => {
-        setRef(element, menuDiv);
+        menuDiv.current = element as HTMLDivElement;
```
Are you not able to just do?

```
ref={menuDiv}
```

dmcrorieGIT replied:
It doesn't look like I can. When I try, it gives me an error since I'm using a `useRef`, which initializes with `undefined`. If I use `createRef` instead, this works fine here, but since we're using keydown listeners the state of the ref is always `undefined`, and so the desired behaviour is lost 🤔

dmcrorieGIT replied:
Just to clarify, it technically "works" using `ref={menuDiv}`, but the typescript linter blows up about it 😅 

![Screen Shot 2020-09-15 at 1 43 01 PM](https://user-images.githubusercontent.com/34727471/93262632-67921b80-f759-11ea-80ea-5a127ece3092.png)

seanhealy replied:
I _think_ if you refine your ref as follows it should clear this up.

```ts
const menuDiv = useRef() as MutableRefObject<HTMLDivElement>;
```

I suspect what is happening is typescript is unhappy that you aren't handing an undefined case that in practical terms won't come up.

dmcrorieGIT replied:
Oh yeah, that worked!! 🤩

seanhealy reviewed (changes requested):
Looking good. Few other thoughts for you here. 😄

dmcrorieGIT commented:
@seanhealy Thanks for the second round of feedback!! I addressed your comments. Could we do another round?? 😅 

![another-thor](https://user-images.githubusercontent.com/34727471/93262457-1b46db80-f759-11ea-945c-adfa2fb7f71c.gif)

seanhealy commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -99,15 +101,45 @@ export function Menu({
+      direction: "up" | "down",
+    ) {
+      const itemDiv = menuDivElement.querySelector(
+        `button.${classnames(styles.option)}:nth-child(${highlightedIndex +
```
Should just be able to do:

```suggestion
        `button.${styles.option}:nth-child(${highlightedIndex +
```

dmcrorieGIT replied:
Oh nice! took out `classnames`

seanhealy commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -99,15 +101,45 @@ export function Menu({
+        `button.${classnames(styles.option)}:nth-child(${highlightedIndex +
+          1})`,
+      ) as HTMLButtonElement;
+      if (!itemDiv) {
+        return;
+      }
```
Just personal preference for a guard statement.

```suggestion
      if (!itemDiv) return;
```

dmcrorieGIT replied:
Done!

seanhealy commented on packages/components/src/Autocomplete/Menu.tsx:
```diff
@@ -99,15 +101,45 @@ export function Menu({
+        return;
+      }
+      const menuTop = menuDivElement.getBoundingClientRect().top;
+      const itemTop = itemDiv.getBoundingClientRect().top;
+      const itemHeight = itemDiv.getBoundingClientRect().height;
```
```suggestion
      const {top: itemTop, height: itemHeight} = itemDiv.getBoundingClientRect();
```

seanhealy replied:
I'm so so on this suggestion. It's _slightly_ better as it's not re-calling `itemDiv.getBoundingClientRect` for each variable but it may not be as readable

dmcrorieGIT replied:
I like this better than getting the bounding rect for each element, so I'm going to use this change 😎

dmcrorieGIT commented:
@seanhealy More changes addressed!
